src/fdentry.rs

```
@@ -0,0 +1,120 @@
+use super::host;
+use crate::sys::fdentry_impl;
+
+use std::fs;
+use std::io;
+use std::mem::ManuallyDrop;
+use std::path::PathBuf;
+
+#[derive(Debug)]
+pub enum Descriptor {
+    File(ManuallyDrop<fs::File>),
+    Stdin,
+    Stdout,
+    Stderr,
+}
+
+#[derive(Debug)]
+pub struct FdObject {
+    pub file_type: host::__wasi_filetype_t,
+    pub descriptor: Descriptor,
+    pub needs_close: bool,
+    // TODO: directories
+}
+
+#[derive(Debug)]
+pub struct FdEntry {
+    pub fd_object: FdObject,
+    pub rights_base: host::__wasi_rights_t,
+    pub rights_inheriting: host::__wasi_rights_t,
+    pub preopen_path: Option<PathBuf>,
+}
+
+impl Drop for FdObject {
+    fn drop(&mut self) {
+        if let Descriptor::File(f) = &mut self.descriptor {
+            if self.needs_close {
+                unsafe { ManuallyDrop::drop(f) }; // this drops the `file`
+            }
+        }
+    }
+}
+
+impl FdEntry {
+    pub fn from(file: fs::File) -> Self {
+        let (file_type, rights_base, rights_inheriting) =
+            fdentry_impl::determine_type_and_access_rights(&file)
+                .expect("could determine type and access rights");
+        let file = ManuallyDrop::new(file);
+
+        Self {
+            fd_object: FdObject {
+                file_type,
+                descriptor: Descriptor::File(file),
+                needs_close: true,
+            },
+            rights_base,
+            rights_inheriting,
+            preopen_path: None,
+        }
+    }
+
+    pub fn duplicate(file: &fs::File) -> Self {
+        let file = file.try_clone().expect("could duplicate file");
```

I haven't seen everything else in this PR yet but this seems like it may want to not use `.expect` and instead return an error?

---

I agree we should return an error here, but, for the moment, I'd leave `WasiCtx` and `FdEntry` as is so that I won't break the public interface of the crate (`WasiCtx` is used in `wasmtime-wasi` crate inside `wasmtime`). What do you reckon?

---

OK, change of plans. Since I'll still need to bump up Rust's version in `wasmtime` to the latest stable (`1.36.0`) due to the use of vectored IO, I've removed the `expect` as suggested. Plus, I've gone ahead and weeded out all uses of `expect` that I could find and replaced them with `Result`s instead.